lib/charms/opensearch/v0/opensearch_locking.py

```diff
@@ -25,78 +25,240 @@
+        if acquired:
+            logger.debug("[Node lock] Acquired via peer databag")
+        else:
+            logger.debug(
+                f"[Node lock] Not acquired. Unit with peer databag lock: {self._unit_with_lock}"
+            )
+        return acquired
+
+    def release(self):
+        """Release lock for this unit."""
+        if not self._relation:
             return
+        self._relation.data[self._charm.unit].pop("lock-requested", None)
+        if self._charm.unit.is_leader():
+            logger.debug("[Node lock] Released peer lock as leader unit")
+            # A separate relation-changed event won't get fired
+            self._on_peer_relation_changed()
+
+    def _unit_requested_lock(self, unit: ops.Unit):
+        """Whether unit requested lock."""
+        assert self._relation
+        value = self._relation.data[unit].get("lock-requested")
+        if not value:
+            return False
+        value = json.loads(value)
+        if not isinstance(value, bool):
+            raise ValueError
+        return value
+
+    @property
+    def _unit_with_lock(self):
+        if self._relation:
+            return self._relation.data[self._charm.app].get("unit-with-lock")
+
+    @_unit_with_lock.setter
+    def _unit_with_lock(self, value: str):
+        assert self._relation
+        self._relation.data[self._charm.app]["unit-with-lock"] = value
+
+    @_unit_with_lock.deleter
+    def _unit_with_lock(self):
+        assert self._relation
+        self._relation.data[self._charm.app].pop("unit-with-lock", None)
+
+    @property
+    def _relation(self):
+        # Use property instead of `self._relation =` in `__init__()` because of ops Harness unit
+        # tests
+        return self._charm.model.get_relation(self._NAME)
+
+    def _on_peer_relation_changed(self, _=None):
+        """Grant & release lock."""
+        if not self._charm.unit.is_leader():
+            return
+        assert self._relation
+        if self._unit_with_lock and self._unit_requested_lock(
+            self._charm.model.get_unit(self._unit_with_lock)
+        ):
+            # Lock still in use, do not release
+            logger.debug("[Node lock] (leader) lock still in use")
+            return
+        # TODO: adjust which unit gets priority on lock?
+        for unit in (*self._relation.units, self._charm.unit):
+            if self._unit_requested_lock(unit):
+                self._unit_with_lock = unit.name
+                logger.debug(f"[Node lock] (leader) granted peer lock to {unit.name=}")
+                break
+        else:
+            logger.debug("[Node lock] (leader) cleared peer lock")
+            del self._unit_with_lock
+
 
-        # we check first on the peer data bag if the lock is already acquired
-        if self._is_lock_in_peer_data():
-            raise OpenSearchOpsLockAlreadyAcquiredError("Another unit is being removed.")
+class OpenSearchNodeLock(ops.Object):
```

Comment:
I think we should add a "token" to the peer relation once the leader successfully acquires the node lock the first time. This "token" would force all the units to never use the peer relation again.

If we have a cluster that got so disconnected it needs to fallback to the peer relation, then we have waaay more problems to deal before dealing with the lock itself. And if we manage to recover the cluster, then we have a functional cluster again.

Reply:
@Mehdi-Bendriss , wdyt?

Reply:
Yeah I think the only case where this might not work is on a full cluster crash of the units (not the services, because systemd services restart automatically) - which would essentially be treated the same as a "new cluster" (lock wise)